seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.connectors.seatunnel.tdengine.source;
+
+import org.apache.seatunnel.api.source.Boundedness;
+import org.apache.seatunnel.api.source.Collector;
+import org.apache.seatunnel.api.source.SourceReader;
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+import org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig;
+import org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorErrorCode;
+import org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException;
+
+import com.google.common.collect.Sets;
+import com.taosdata.jdbc.TSDBDriver;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+
+@Slf4j
+public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengineSourceSplit> {
+
+    private static final long THREAD_WAIT_TIME = 500L;
+
+    private final TDengineSourceConfig config;
+
+    private final Set<TDengineSourceSplit> sourceSplits;
+
+    private final Context context;
+
+    private Connection conn;
+
+    public TDengineSourceReader(TDengineSourceConfig config, SourceReader.Context readerContext) {
+        this.config = config;
+        this.sourceSplits = Sets.newHashSet();
+        this.context = readerContext;
+    }
+
+    @Override
+    public void pollNext(Collector<SeaTunnelRow> collector) throws InterruptedException {
+        if (sourceSplits.isEmpty()) {
+            Thread.sleep(THREAD_WAIT_TIME);
+            return;
+        }
+        sourceSplits.forEach(split -> {
```

Comment:
During the read split and changing state(sourceSplits), you need to obtain CheckpointLock, otherwise it cannot be guaranteed to be exactly-once.

Reply:
like
```
synchronized (output.getCheckpointLock()) {
  sourceSplits.forEach(split -> {
       read(split, collector);
  }
}
```
Or refer to JdbcSourceReader to use a smaller lock granularity

Reply:
The implementation in JDBC connector `org.apache.seatunnel.connectors.seatunnel.jdbc.source.JdbcSourceReader#pollNext`  is as same as `synchronized (output.getCheckpointLock())`. Where can I refer to using a smaller lock granularity?